core/src/main/scala/kafka/controller/PartitionStateMachine.scala

```diff
@@ -135,207 +146,285 @@ class PartitionStateMachine(controller: KafkaController, stateChangeLogger: Stat
+  private def doElectLeaderForPartitions(partitions: Seq[TopicAndPartition], partitionLeaderElectionStrategy: PartitionLeaderElectionStrategy):
+  (Seq[TopicAndPartition],
+    Seq[TopicAndPartition],
+    Map[TopicAndPartition, Exception]) = {
+    val getDataResponses = try {
+      zkUtils.getTopicPartitionStatesRaw(partitions)
+    } catch {
+      case e: Exception =>
+        return (Seq.empty, Seq.empty, partitions.map(_ -> e).toMap)
+    }
+    val failedElections = mutable.Map.empty[TopicAndPartition, Exception]
+    val leaderIsrAndControllerEpochPerPartition = mutable.Buffer.empty[(TopicAndPartition, LeaderIsrAndControllerEpoch)]
+    getDataResponses.foreach { getDataResponse =>
+      val partition = getDataResponse.ctx.asInstanceOf[TopicAndPartition]
+      val currState = partitionState(partition)
+      if (Code.get(getDataResponse.rc) == Code.OK) {
+        val leaderIsrAndControllerEpochOpt = TopicPartitionStateZNode.decode(getDataResponse.data, getDataResponse.stat)
+        if (leaderIsrAndControllerEpochOpt.isEmpty) {
+          val exception = new StateChangeFailedException(s"LeaderAndIsr information doesn't exist for partition $partition in $currState state")
+          failedElections.put(partition, exception)
         }
+        leaderIsrAndControllerEpochPerPartition += partition -> leaderIsrAndControllerEpochOpt.get
+      } else if (Code.get(getDataResponse.rc) == Code.NONODE) {
+        val exception = new StateChangeFailedException(s"LeaderAndIsr information doesn't exist for partition $partition in $currState state")
+        failedElections.put(partition, exception)
+      } else {
+        failedElections.put(partition, KeeperException.create(Code.get(getDataResponse.rc)))
+      }
+    }
+    val (invalidPartitionsForElection, validPartitionsForElection) = leaderIsrAndControllerEpochPerPartition.partition { case (partition, leaderIsrAndControllerEpoch) =>
+      leaderIsrAndControllerEpoch.controllerEpoch > controllerContext.epoch
+    }
+    invalidPartitionsForElection.foreach { case (partition, leaderIsrAndControllerEpoch) =>
+      val failMsg = s"aborted leader election for partition $partition since the LeaderAndIsr path was " +
+        s"already written by another controller. This probably means that the current controller $controllerId went through " +
+        s"a soft failure and another controller was elected with epoch ${leaderIsrAndControllerEpoch.controllerEpoch}."
+      failedElections.put(partition, new StateChangeFailedException(failMsg))
+    }
+    if (validPartitionsForElection.isEmpty) {
+      return (Seq.empty, Seq.empty, failedElections.toMap)
+    }
+    val shuttingDownBrokers  = controllerContext.shuttingDownBrokerIds.toSet
+    val (partitionsWithoutLeaders, partitionsWithLeaders) = partitionLeaderElectionStrategy match {
+      case OfflinePartitionLeaderElectionStrategy =>
+        leaderForOffline(validPartitionsForElection).partition(_._2.isEmpty)
```

Comment:
Could we use case inside partition to avoid unnamed reference _._2? There are a few other places like that.